src/ParametricExpression.jl

```diff
@@ -0,0 +1,326 @@
+        n = constructorof(typeof(t))(; val=zero(T))
+        n.constant = false
+        n.is_parameter = true
+        n.parameter = t.parameter
+        n
+    end
+    return out
+end
+function leaf_hash(h::UInt, t::ParametricNode)
+    if t.constant
+        hash((:constant, t.val), h)
+    else
+        if t.is_parameter
+            hash((:parameter, t.parameter), h)
+        else
+            hash((:feature, t.feature), h)
+        end
+    end
+end
+function leaf_equal(a::ParametricNode, b::ParametricNode)
+    if a.constant
+        return b.constant && a.val == b.val
+    else
+        if a.is_parameter
+            return b.is_parameter && a.parameter == b.parameter
+        else
+            return a.feature == b.feature
+        end
+    end
+end
+###############################################################################
+
+###############################################################################
+# Abstract expression interface ###############################################
+###############################################################################
+function get_tree(ex::ParametricExpression)
+    return ex.tree
+end
+function get_operators(ex::ParametricExpression, operators)
+    return operators === nothing ? ex.metadata.operators : operators
+end
+function get_variable_names(ex::ParametricExpression, variable_names)
+    return variable_names === nothing ? ex.metadata.variable_names : variable_names
+end
+@inline _copy_with_nothing(x) = copy(x)
+@inline _copy_with_nothing(::Nothing) = nothing
+function Base.copy(ex::ParametricExpression; break_sharing::Val=Val(false))
+    return ParametricExpression(
+        copy(ex.tree; break_sharing=break_sharing);
+        operators=_copy_with_nothing(ex.metadata.operators),
+        variable_names=_copy_with_nothing(ex.metadata.variable_names),
+        parameters=_copy_with_nothing(ex.metadata.parameters),
+        parameter_names=_copy_with_nothing(ex.metadata.parameter_names),
+    )
+end
+function Base.hash(ex::ParametricExpression, h::UInt)
+    return hash(ex.tree, hash(ex.metadata, h))
+end
+function Base.:(==)(x::ParametricExpression, y::ParametricExpression)
+    return x.tree == y.tree && x.metadata == y.metadata
+end
+###############################################################################
+
+###############################################################################
+# Extra utilities for parametric-specific behavior ############################
+###############################################################################
+## As explained in AbstractExpressionNode, we can implement custom behavior for
+## the parametric expression by implementing the following methods:
+# - `count_nodes`
+# - `count_constants`
+# - `count_depth`
+# - `index_constants`
+# - `has_operators`
+# - `has_constants`
+# - `get_constants`
+# - `set_constants!`
+# - `string_tree`
+# - `max_feature`
+# - `eval_tree_array`
+# - `eval_grad_tree_array`
+# - `_grad_evaluator`
```

Comment:
These should live in the docs instead

Reply:
Would also be nice to have https://github.com/rafaqz/Interfaces.jl added so we can check all of this stuff.

Reply:
Ok added Interfaces.jl